http/src/v0/block.rs

```diff
@@ -102,8 +81,106 @@ pub fn put<T: IpfsTypes>(
     path!("block" / "put")
         .and(with_ipfs(ipfs))
         .and(query::<PutQuery>())
-        .and(multipart::form())
-        .and_then(put_query)
+        .and(warp::header::<Mime>("content-type"))
+        .and(warp::body::stream())
+        .and_then(inner_put)
+}
+
+async fn inner_put<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    opts: PutQuery,
+    mime: Mime,
+    body: impl Stream<Item = Result<impl Buf, warp::Error>> + Unpin,
+) -> Result<impl Reply, Rejection> {
+    let boundary = mime.get_param("boundary").map(|v| v.to_string()).unwrap();
+    let mut stream = MultipartStream::new(boundary, body.map_ok(|mut buf| buf.to_bytes()));
+
+    // store the first good field here; optimally this would just be an Option but couldn't figure
+    // out a way to handle the "field matched", "field not matched" cases while supporting empty
+    // fields.
+    let mut buffer = Vec::new();
+    let mut matched = false;
+
+    loop {
+        match stream.try_next().await {
+            Ok(Some(mut field)) => {
+                // [ipfs http api] says we should expect a "data" but examples use "file" as the
+                // form field name. newer conformance tests also use former, older latter.
+                //
+                // [ipfs http api]: https://docs.ipfs.io/reference/http/api/#api-v0-block-put
+
+                let name = field
+                    .name()
+                    .map_err(|_| StringError::from("invalid field name"))?;
+
+                let mut target = if name != "data" {
+                    if name != "file" {
+                        log::warn!("block/put: discarding part with name {:?}", name);
+                        None
+                    } else {
+                        log::warn!("block/put: processing multipart with name {:?}", name);
+                        Some(&mut buffer)
+                    }
+                } else {
+                    Some(&mut buffer)
+                };
```

Comment:
nit: I'd prefer a `match name` with `"file"`, `"data"` and `_` branches, for extra clarity